File: leetcode/61-rotate-list/main.go
package main

import (
	"fmt"
)

/**
 * Definition for singly-linked list.
 */
type ListNode struct {
	Val  int
	Next *ListNode
}

func rotateRight(head *ListNode, k int) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}

	p := head
	// Finding out the length of the list
	var length = 1
	for p.Next != nil {
		length++
		p = p.Next
	}
	p.Next = head

	for i := length - k%length; i > 1; i-- {
		head = head.Next
	}

	p = head.Next
	head.Next = nil

	return p
}

func main() {
	list := &ListNode{
		Val: 1,
		Next: &ListNode{
			Val: 2,
			Next: &ListNode{
				Val: 3,
				Next: &ListNode{
					Val: 4,
					Next: &ListNode{
						Val: 5,
					},
				},
			},
		},
	}

	fmt.Println(rotateRight(list, 2))
}

File: leetcode/61-rotate-list/main.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x, next = None):
        self.val = x
        self.next = next


class Solution:
    def rotateRight(self, head: ListNode, k: int) -> ListNode:
        if head is None:
            return None

        length, tail = 1, head
        while tail.next:
            length += 1
            tail = tail.next
        tail.next = head
        steps = length - k % length
        while steps:
            steps -= 1
            tail = tail.next
        head = tail.next
        tail.next = None

        return head


solution = Solution()
head = ListNode(0, ListNode(1, ListNode(2)))
print(solution.rotateRight(head, 4).val)
